### PR TITLE
Add name method to execution plan

### LINKED ISSUE #9558: Add `name` method to `ExecutionPlan` trait
### Is your feature request related to a problem or challenge?

In the telemetry we produce on query executions we sometimes need to know the name of `ExecutionPlan` nodes (i.e. `ParquetExec`).  Right now we do something like the following to extract the name:

```
// Get the one-line representation of the ExecutionPlan, something like this:
//   ParquetExec: file_groups=[...], ...
let mut buf = String::new();
write!(&mut buf, "{}", displayable(plan).one_line()).map_err(|e| {
    DataFusionError::Internal(format!("Error while collecting metrics: {e}"))
 })?;
 let node_type = match buf.split_once(':') {
    None => {
        warn!("execution plan has unexpected display format: {buf}");
           return Ok(true);
    }
   Some((name, _)) => name.to_string(),
 };
```

From what I have seen the name of `ExecutionPlan` is always known at compile time so I think this is unnecessary work and I am hoping to provide a simpler way to get access to this information. 

### Describe the solution you'd like

Could we add something like the following to `ExecutionPlan`

```
pub trait ExecutionPlan: Debug + DisplayAs + Send + Sync {
    const NAME: &'static str;

    fn name(&self) -> &'static str {
        return Self::NAME;
    }

impl ExecutionPlan for ParquetExec {
    const NAME = "ParquetExec";

    // Rest of `ParquetExec` is unchanged
    ....
}
```

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by matthewmturner**
## Which issue does this PR close?

Closes #9558

### CODE REVIEW

**Review by comphead — COMMENTED** [review]
Id say its LGTM, altough I would probably more incline to use just a default method with`std::any::type_name::<Self>();` so the code will be cleaner, but looks like this option was discussed

**Comment by matthewmturner** [comment]
> Id say its LGTM, altough I would probably more incline to use just a default method with`std::any::type_name::<Self>();` so the code will be cleaner, but looks like this option was discussed

After rethinking I think that that option is indeed a better default.  I will update to that.  FYI @SteveLauC

**Comment by matthewmturner** [comment]
@comphead @SteveLauC I have updated the default impl.  Let me know if any more concerns.

File: datafusion/physical-plan/src/lib.rs [hunk 1]
```diff
@@ -113,6 +113,15 @@ pub use datafusion_execution::{RecordBatchStream, SendableRecordBatchStream};
 /// [`required_input_distribution`]: ExecutionPlan::required_input_distribution
 /// [`required_input_ordering`]: ExecutionPlan::required_input_ordering
 pub trait ExecutionPlan: Debug + DisplayAs + Send + Sync {
+    /// Short name for the ExecutionPlan, such as 'ParquetExec'.
+    fn name(&self) -> &'static str {
+        let full_name = std::any::type_name::<Self>();
+        let maybe_start_idx = full_name.rfind(":");
```
Comment (comphead) [inline datafusion/physical-plan/src/lib.rs:119]: thanks @matthewmturner would you mind adding a unit test?

File: datafusion/physical-plan/src/lib.rs [hunk 2]
```diff
@@ -113,6 +113,15 @@ pub use datafusion_execution::{RecordBatchStream, SendableRecordBatchStream};
 /// [`required_input_distribution`]: ExecutionPlan::required_input_distribution
 /// [`required_input_ordering`]: ExecutionPlan::required_input_ordering
 pub trait ExecutionPlan: Debug + DisplayAs + Send + Sync {
+    /// Short name for the ExecutionPlan, such as 'ParquetExec'.
+    fn name(&self) -> &'static str {
+        let full_name = std::any::type_name::<Self>();
```
Comment (SteveLauC) [inline datafusion/physical-plan/src/lib.rs:118]: Just realized that users can implement their own implementations if they don't want this method to involve any computation, nice!

**Review by comphead — COMMENTED** [review]
thanks @matthewmturner I think we really close.
Reg to test, wondering can it be smaller? I'd say we dont need the full implementations for trait methods, just stubs because you testing out just `.name()` field and its overrides?

Instead of implementations you can put macros like `todo()!`, 'unreacheable()!', 'not_implemented()!'

**Review by comphead — APPROVED** [review]
Thanks @matthewmturner lgtm
1 more thing is probably we need to ensure the name is unique? if implementation will be separated across different crates and users can develop their own implementations it can possibly be that the name will not be unique for 2 different structs

**Comment by matthewmturner** [comment]
@comphead thats a good / interesting point.  what comes to mind is perhaps datafusion statically defining / guaranteeing uniqueness of builtin exec names (perhaps could be automated?) and then users would be required to ensure uniqueness of their own execs. im not sure we would want to add some runtime check for that.

i think this could be a follow on PR but let me know if you think otherwise.

**Comment by comphead** [comment]
I'm okay with the followup, as there is only 1 use case for now, but we need to keep this in mind

**Comment by alamb** [comment]
Thanks everyone!

**Pull request merged**